### DISCUSSION TITLE
Graphql Relay Compliant

### DISCUSSION THREAD

**Discussion by jahead** [directus/directus#7772] [Feature Requests]
It would be awesome, if this GraphQL api generation had the ability to produce a compliant API for Relay. I would jump on this project in a heartbeat
see below:
https://relay.dev/docs/guides/graphql-server-specification/

**Comment by benhaynes** [comment]
Is this about changing our API structure, or about generating a specification file (similar to our OAS endpoint)? @rijkvanzanten thoughts?

**Reply by jahead** [reply]
I'm not sure changing your api structure permanently, because relay is opinionated, and Relay in itself can be bit of a task to get your head around.  Maybe a setting?

It definitely requires a change on how you map certain relations in the graph. They call the 'join'ing nodes connections. From the spec. 
``` graphql
query MoreRebelShipsQuery {
  rebels {
    name,
    ships(first: 2) {
      edges {
        cursor
        node {
          name
        }
      }
    }
  }
}
```
and we get back
``` graphql
{
  "rebels": {
    "name": "Alliance to Restore the Republic",
    "ships": {
      "edges": [
        {
          "cursor": "YXJyYXljb25uZWN0aW9uOjA=",
          "node": {
            "name": "X-Wing"
          }
        },
        {
          "cursor": "YXJyYXljb25uZWN0aW9uOjE=",
          "node": {
            "name": "Y-Wing"
          }
        }
      ]
    }
  }
}
```

Basically they have a formal way to handle pagination and search traversing the graph and this 'opinion' is translated to their react relay packages. The documentation/videoes explain why this is 'good' and how relay abuses this to do some cool things. It will explain it way better then I would. 

This connection stuff is probably the biggest change. 

The other requirement is pretty somewhat straight forward to comply with, and I think this is becoming more common in 'standard' graphql. 
Global Object Identification

``` graphql
query RebelsRefetchQuery {
  node(id: "RmFjdGlvbjox") {
    id
    ... on Faction {
      name
    }
  }
}
```
returns 
``` graphql
{
  "node": {
    "id": "RmFjdGlvbjox",
    "name": "Alliance to Restore the Republic"
  }
}
```
https://relay.dev/docs/guides/graphql-server-specification/#object-identification
The graphql website goes into the 'why' but relay assumes your graph has a 'node' it can use to re-fetch an object, 
https://graphql.org/learn/global-object-identification/

**Reply by rijkvanzanten** [reply]
I'm down to look into the possibility of supporting the relay structure, though I wouldn't want to make it the default. Ideally, we can make it an opt-in on the fly through a `?relay` query parameter or something similar. 

We currently rely on [graphql-compose](https://graphql-compose.github.io) to build out the graphql schema, and i saw [this package](https://github.com/graphql-compose/graphql-compose-relay) that can automatically inject the relay specific _stuff_, so maybe that would be an option to support this 🙂

**Reply by jahead** [reply]
How'd you go in your investigation @rijkvanzanten ?

**Reply by joselcvarela** [reply]
Hey @jahead 
As you can see here https://docs.directus.io/contributing/introduction/#feature-requests we use a rule called 80/20.
Right now, this improvement only has one upvote (yours) so as you may understand, it is not a priority.
But we are open to receive new PRs and if they fit our standards we could merge.
Since @rijkvanzanten already gave some lights how this could be done, we will be happy if you decide to open a PR for this 🙂

**Reply by jahead** [reply]
Ah! awesome, was wondering where the opinion landed on this, thought it was still in investigation stage. If I find the time to do a PR. I'd be willing. We are still finding a use for this product as fueling some internal API pro-typing. And you all have a lot on plate so it's understandable.

Happy developing :)

**Reply by joselcvarela** [reply]
Thank you for your understanding ❤
But if this is a improvement you would want for us to make it, we also have this https://docs.directus.io/getting-started/support/#commissioned-features

**Reply by rijkvanzanten** [reply]
@jahead Oh I'm not actively working on this 😬 Apologies if that wasn't clear from the message. What I meant earlier is that I'm more than down to having the option in the API in general, and that we might be able to rely on that other package to make it happen. That being said, it's not something we can implement immediately, as other things have a higher priority 🙂

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
